### ISSUE TITLE
"stan: invalid start time" exception

### ISSUE THREAD

**Issue by zshamrock** [nats-io/java-nats-streaming#22]
When running the following example https://gist.github.com/zshamrock/16215ecdccfac107212fabcc4af88bc2 (Subscriber.java), I get the corresponding exception:

```
Exception in thread "main" java.io.IOException: stan: invalid start time
    at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:464)
    at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:402)
```

(passing `session.ceba04ba-aeff-414c-8912-1eb337db0ddc 1473261548683` as program arguments)

Although if I do run in debug (so, there is a delay in between) then the subscriber starts without any errors. Although depends on the delay I get different messages (the start sequence number differs between the runs, i.e. in one run it starts from seq num 54, in another from seq num 17, in another one from seq num 3, etc.).

You can reproduce it locally using the same classes (Publisher, and Subscriber) from the Gist above.

Any idea what it can be?

**Comment by mcqueary** [comment]
Can you verify that the timestamp you're passing to the Subscriber is >= the oldest event that `nats-streaming-server` has stored? If the timestamp you supply is older than the first stored event, you'll get the error indicated.

**Comment by mcqueary** [comment]
As long as I enter a start timestamp that is ~10ms > the Publisher's start timestamp, there is no exception. Please bear in mind that regardless of what time the publisher starts:

- It does take a small but measurable amount of time for the connection to be created and the first message to be published
- messages are timestamped at the server, after they have been received, and just before they are stored
- clock resolution on most systems is still no better than ~10ms. 

So in an artificial example, you need to budget your try things this way:

- Start `nats-streaming-server`
- Window 1: start `Publisher` (noting timestamp)
- wait a few seconds
- Window 2: start `Subscriber`, supplying a timestamp ~10-100ms > `Publisher`'s start timestamp

Also, as a side note, you don't need to create the NATS connection yourself if you don't want to. You just need to call `cf.setNatsUrl("nats://localhost:4333");` on the STAN `ConnectionFactory` before you call `cf.createConnection(...)`.

**Comment by zshamrock** [comment]
Yes, I think this is the case. I also saw similar errors when I try to use the very big seq number (greater than total number of messages available) or delta (N seconds ago) when no messages are available.

2 questions:
- can the errors be then more user-friendly, if possible, explaining, why the value is invalid
- on bigger scope - why it is an error at all? Is it a technical limitations/architecture of NATS streaming server? So, it puts more efforts on the client to keep track of the timestamp of the first message, the latest seq number, timestamp of the latest message (so N time unit ago will not fail). Is it done for a reason?

**Comment by mcqueary** [comment]
Aliaksandr, you raise a good point(s), both about the error and the fact that it is an error at all. We discussed this internally during implementation and didn't come to a consensus, so your feedback is useful.

I think we can at least make the error message more friendly.  As to whether we consider it an error, or just start with the earliest timestamp we DO have, we'll keep discussing.

**Comment by zshamrock** [comment]
A few more thoughts/examples on the "error":

(`session-replay` is the tool I wrote, to fetch messages/subscribe to the subject):

### Start at seq
$ session-replay -sid ba19263a-e347-4c29-85ed-1ffcfceb9282
```
Sep 09, 2016 5:01:31 PM session.replay.PubSub createConnection
INFO: Connecting to nats://localhost:4333 for cluster 'events' with client ID 'c2d28a66-8b93-4f72-a0ff-1fe041ba00cf'
Sep 09, 2016 5:01:31 PM session.replay.Subscriber subscribe
INFO: Subscribing messages from session.ba19263a-e347-4c29-85ed-1ffcfceb9282
8lbKJETUf8@1473439759091
w7Zektf5ag@1473439760190
JmW9MjjZZU@1473439761205
G2UM1RCgvM@1473439762216
oewQbrFszd@1473439763227
jTGzKcf25c@1473439764237
Ylg7rGhI9y@1473439765247
yfrGd6DgV7@1473439766257
OVmOB5Q8Ih@1473439767264
mS0Z3cYA0Z@1473439768275
JcjwUpNoOL@1473439769287
7LkImliHNs@1473439770296
```
So, there are 12 messages in the subject/channel.
Asking for seq 12, returns the latest one:
```
session-replay -sid ba19263a-e347-4c29-85ed-1ffcfceb9282 -seq 12
Sep 09, 2016 5:04:19 PM session.replay.PubSub createConnection
INFO: Connecting to nats://localhost:4333 for cluster 'events' with client ID '95d1b943-e829-4e67-b664-e5150a9462a1'
Sep 09, 2016 5:04:19 PM session.replay.Subscriber subscribe
INFO: Subscribing messages from session.ba19263a-e347-4c29-85ed-1ffcfceb9282
7LkImliHNs@1473439770296
```
Although, if I ask for seq 13, it fails (as "expected"):
```
session-replay -sid ba19263a-e347-4c29-85ed-1ffcfceb9282 -seq 13
Sep 09, 2016 5:05:28 PM session.replay.PubSub createConnection
INFO: Connecting to nats://localhost:4333 for cluster 'events' with client ID '54da0ef7-6f69-4128-9a77-354d80ba1b98'
Sep 09, 2016 5:05:28 PM session.replay.Subscriber subscribe
INFO: Subscribing messages from session.ba19263a-e347-4c29-85ed-1ffcfceb9282
Exception in thread "main" java.io.IOException: stan: invalid start sequence
        at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:464)
        at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:402)
        at session.replay.Subscriber.subscribe(Subscriber.java:40)
        at session.replay.SessionReplayConsoleClient.run(SessionReplayConsoleClient.java:108)
        at session.replay.SessionReplayConsoleClient.main(SessionReplayConsoleClient.java:82)
```
It is difficult to come with a real use case where asking for specific *unknown* seq number will be required. As I imagine when someone have to ask for specific seq number, I assume he knows this seq number exists. So, not very much a problem. So in this specific (sequence number) use case, it might be reasonable to throw an exception.
Although one use case can be if someone subscribes in advance, and for some reason wants to start processing messages from 10'000, so will wait until this message is in the subject (again how realistic it is).

### Start at time delta
This use case looks like a more common/typical one. Assuming there is a server running, when I connect I am interested from the messages starting from, lets say, 10 seconds ago. If the publisher is running, and there is no long pause between the messages (smaller than 10 seconds, it will never fail), although I can imagine when messages are not coming that often (then another question will be, whether it is a stream of events, and whether NATS streaming is the right tool, as it assumes continuous stream of events).
Ex. (publisher is running in the background `INFO: Publishing messages to session.1864bc98-54c2-4e19-8be5-e962eb677bd7 at 1473441331219 ms`):
```
session-replay -sid 1864bc98-54c2-4e19-8be5-e962eb677bd7 -ago 5                                                                            
Sep 09, 2016 5:16:08 PM session.replay.PubSub createConnection
INFO: Connecting to nats://localhost:4333 for cluster 'events' with client ID '09d0ae42-8150-4136-9bab-728036b970c5'
Sep 09, 2016 5:16:08 PM session.replay.Subscriber subscribe
INFO: Subscribing messages from session.1864bc98-54c2-4e19-8be5-e962eb677bd7
DIOHfP6LRz@1473441363817
kkyfCbc3Fe@1473441364822
92UXbRPa0L@1473441365832
cevYPiQHvW@1473441366837
V2zaEQKuAv@1473441367847
ru1FhjTis1@1473441368855
xtKYSEKgQi@1473441369862
```
Works fine.
Now stop the publisher, and wait more than 5 seconds:
```
session-replay -sid 1864bc98-54c2-4e19-8be5-e962eb677bd7 -ago 5
Sep 09, 2016 5:17:16 PM session.replay.PubSub createConnection
INFO: Connecting to nats://localhost:4333 for cluster 'events' with client ID '5a25a6da-9881-4be4-9c1c-c319806595bc'
Sep 09, 2016 5:17:16 PM session.replay.Subscriber subscribe
INFO: Subscribing messages from session.1864bc98-54c2-4e19-8be5-e962eb677bd7
Exception in thread "main" java.io.IOException: stan: invalid start time
        at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:464)
        at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:402)
        at session.replay.Subscriber.subscribe(Subscriber.java:40)
        at session.replay.SessionReplayConsoleClient.run(SessionReplayConsoleClient.java:108)
        at session.replay.SessionReplayConsoleClient.main(SessionReplayConsoleClient.java:82)
```
Even more the challenge to come with the correct time unit for the "stopped" stream, as `ago` must fit into the `[start, end]` interval of the stream.
Again if the stream is continuous, and someone ask for N seconds ago, it should always work. So, again, it might a be correct behavior.

### Start at time
(`session-replay` doesn't support this, yet)
But, it might be similar to the *Start at seq* use case, if someone asks for the start at time, he might know the messages are there. In our use case, we actually need all the messages, so we just ask for `deliverAllAvailable()` or we know the stream is running, just start from a couple of seconds ago, to see how events look like.

---
So, in summary, it might be the right behavior to throw an exception (just more user friendly messages will help for sure). Until someone can come with a reasonable use case why it should not.

Thoughts?

**Comment by zshamrock** [comment]
One more thing, back to my original statement from the issue:

> Although if I do run in debug (so, there is a delay in between) then the subscriber starts without any errors. Although depends on the delay I get different messages (the start sequence number differs between the runs, i.e. in one run it starts from seq num 54, in another from seq num 17, in another one from seq num 3, etc.).

I've just tried it one more time, and it is still the case. You can try to reproduce it itself:
1. Run `Publisher` for a few secs, remember session/subject id and timestamp. Stop it.
2. Run `Subscriber` pass the recorded subject and timestamp, it will fail with the exception in the topic
3. Run `Subscriber`, put a breakpoint (at least I have a couple around inside NATS streaming Java client), move slowly from one breakpoint to another. It works, you get messages back. **Although you still subscribe from the same timestamp as in the previous step**. So why it works now? Also it returns back on every run different start message seq. Is there somewhere race condition in the Java client?

Some logs:
### Publish
```
Sep 09, 2016 7:34:19 PM com.akazlou.Publisher start
INFO: Publishing messages to session.01226370-4d80-4b7f-84f6-08892c826d6e at 1473442459687 ms
Sep 09, 2016 7:34:19 PM com.akazlou.Publisher start
INFO: Publishing q3qMaUBtxJ@1473442459832 message
Sep 09, 2016 7:34:20 PM com.akazlou.Publisher start
INFO: Publishing qF2hbargdH@1473442460882 message
Sep 09, 2016 7:34:21 PM com.akazlou.Publisher start
INFO: Publishing zQelDG1a5A@1473442461895 message
Sep 09, 2016 7:34:22 PM com.akazlou.Publisher start
INFO: Publishing o3Ug4hWbUt@1473442462906 message
Sep 09, 2016 7:34:23 PM com.akazlou.Publisher start
INFO: Publishing eJDjeSgUBy@1473442463914 message
Sep 09, 2016 7:34:24 PM com.akazlou.Publisher start
INFO: Publishing 9lMyNoIrta@1473442464922 message
Sep 09, 2016 7:34:25 PM com.akazlou.Publisher start
INFO: Publishing ix8lEjxSyz@1473442465933 message
Sep 09, 2016 7:34:26 PM com.akazlou.Publisher start
INFO: Publishing LJ8ZkbksSL@1473442466941 message
```
### Subscribe without debug/breakpoints
```
Exception in thread "main" java.util.concurrent.TimeoutException: stan: connect request timeout
	at io.nats.stan.ConnectionImpl.connect(ConnectionImpl.java:145)
	at io.nats.stan.ConnectionFactory.createConnection(ConnectionFactory.java:40)
	...
```

### Subscribe with debug/breakpoints
```
Sep 09, 2016 7:36:45 PM com.akazlou.Subscriber start
INFO: Subscribing messages from session.01226370-4d80-4b7f-84f6-08892c826d6e since 1473442459687 ms
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: {Timestamp=1473442461899514458;Sequence=3;Redelivered=false;Subject=session.01226370-4d80-4b7f-84f6-08892c826d6e;Reply=;Payload=<zQelDG1a5A@1473442461895>}
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: zQelDG1a5A@1473442461895
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: {Timestamp=1473442462910395164;Sequence=4;Redelivered=false;Subject=session.01226370-4d80-4b7f-84f6-08892c826d6e;Reply=;Payload=<o3Ug4hWbUt@1473442462906>}
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: o3Ug4hWbUt@1473442462906
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: {Timestamp=1473442463918321561;Sequence=5;Redelivered=false;Subject=session.01226370-4d80-4b7f-84f6-08892c826d6e;Reply=;Payload=<eJDjeSgUBy@1473442463914>}
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: eJDjeSgUBy@1473442463914
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: {Timestamp=1473442464926095571;Sequence=6;Redelivered=false;Subject=session.01226370-4d80-4b7f-84f6-08892c826d6e;Reply=;Payload=<9lMyNoIrta@1473442464922>}
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: 9lMyNoIrta@1473442464922
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: {Timestamp=1473442465938750728;Sequence=7;Redelivered=false;Subject=session.01226370-4d80-4b7f-84f6-08892c826d6e;Reply=;Payload=<ix8lEjxSyz@1473442465933>}
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: ix8lEjxSyz@1473442465933
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: {Timestamp=1473442466945501110;Sequence=8;Redelivered=false;Subject=session.01226370-4d80-4b7f-84f6-08892c826d6e;Reply=;Payload=<LJ8ZkbksSL@1473442466941>}
Sep 09, 2016 7:37:01 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: LJ8ZkbksSL@1473442466941
```
See above, it returns back from the seq=3.

### First message
```
Sep 09, 2016 7:39:52 PM com.akazlou.Subscriber start
INFO: Subscribing messages from session.01226370-4d80-4b7f-84f6-08892c826d6e since 1473442459687 ms
Sep 09, 2016 7:39:52 PM com.akazlou.Subscriber$EventsMessageHandler onMessage
INFO: {Timestamp=1473442459876317138;Sequence=1;Redelivered=false;Subject=session.01226370-4d80-4b7f-84f6-08892c826d6e;Reply=;Payload=<q3qMaUBtxJ@1473442459832>}
```
And as you can see it is "younger" (`Timestamp=1473442459876317138`) than the timestamp I ask from. So, why it works, if I run `Subscriber` in debug mode?

**Comment by kozlovic** [comment]
Note that in `Subscribe without debug/breakpoints` the error is not about invalid subscription, but connect to server failed (`connect request timeout`).
The subscriber (running in debug) receives messages because the given timestamp is >= than the first message and <= than the last.
Note that we are internally discussing if it makes sense at all to return an error both for time and sequence if no message is avail. Will update later.

**Comment by zshamrock** [comment]
Interesting, I've tried once again without debug, and actually it returns 2 different exceptions (I saw `timeout` once, the first time I tried to connect after the publisher was stopped, so probably due to some internals to make subject available?)

All other runs results in the following error:
```
INFO: Subscribing messages from session.439caa70-465b-4db8-a4ec-664c00d6e295 since 1473447580659 ms
Exception in thread "main" java.io.IOException: stan: invalid start time
	at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:464)
	at io.nats.stan.ConnectionImpl.subscribe(ConnectionImpl.java:402)
```

But running in debug I use the very same timestamp value I use without debug mode. So, I've expected it to fail with `Exception in thread "main" java.io.IOException: stan: invalid start time`, as the timestamp I set is definitely older than the first message in the subject (as I do sleep 1 second in the publisher before publishing the 1st message).

**Comment by mcqueary** [comment]
If you can reproduce the timeout exception and stack trace, I'd be interested in seeing it.

**Comment by zshamrock** [comment]
It is difficult, I don't know the exact steps to reproduce it. 

I got one more today, although this time for Publisher. Here is the exception, and all the stack trace I had, and also the log from the stan with -SDV and -DV enabled (not sure if the helps):
```
Exception in thread "main" java.util.concurrent.TimeoutException: stan: connect request timeout
	at io.nats.stan.ConnectionImpl.connect(ConnectionImpl.java:145)
	at io.nats.stan.ConnectionFactory.createConnection(ConnectionFactory.java:40)
	
[2935] 2016/09/10 21:49:45.475171 [TRC] 127.0.0.1:43868 - cid:6 - <<- [MSG _INBOX.L0TUQU9254AE8VEHXM0GHY 1 _INBOX.jjOqOw59rUG6hNIVnFP8u4 0]
[2935] 2016/09/10 21:49:45.477447 [TRC] 127.0.0.1:43868 - cid:6 - ->> [PUB _INBOX.jjOqOw59rUG6hNIVnFP8u4 0]
[2935] 2016/09/10 21:49:45.477471 [TRC] 127.0.0.1:43868 - cid:6 - ->> MSG_PAYLOAD: []
[2935] 2016/09/10 21:49:45.477505 [DBG] 127.0.0.1:43868 - cid:6 - Auto-unsubscribe limit of 1 reached for sid '26'
[2935] 2016/09/10 21:49:45.477517 [TRC] ::1:35155 - cid:2 - <<- [MSG _INBOX.jjOqOw59rUG6hNIVnFP8u4 26 0]
[2935] 2016/09/10 21:49:45.477531 [TRC] ::1:35155 - cid:2 - <-> [DELSUB 26]
[2935] 2016/09/10 21:50:07.377294 [TRC] ::1:35155 - cid:2 - ->> [SUB _INBOX.jjOqOw59rUG6hNIVnFP8ue  27]
[2935] 2016/09/10 21:50:07.377333 [TRC] ::1:35155 - cid:2 - ->> [UNSUB 27 1]
[2935] 2016/09/10 21:50:07.377346 [DBG] ::1:35155 - cid:2 - Deferring actual UNSUB(_INBOX.jjOqOw59rUG6hNIVnFP8ue): 1 max, 0 received
[2935] 2016/09/10 21:50:07.377358 [TRC] ::1:35155 - cid:2 - ->> [PUB _INBOX.OZ6FALEAIAY3KQWKTIYWXY _INBOX.jjOqOw59rUG6hNIVnFP8ue 0]
[2935] 2016/09/10 21:50:07.377368 [TRC] ::1:35155 - cid:2 - ->> MSG_PAYLOAD: []	
```

No, I am not using the latest snapshot, 0.2.2 for the server, and 0.1.2 for the client.

**Comment by sasbury** [comment]
This is an old one, do you still see it in the latest library? Going to close for now, sorry it got stuck during team transitions.
